core/corerepo/gc.go

```diff
@@ -16,26 +16,32 @@ type KeyRemoved struct {
 }
 
 func GarbageCollect(n *core.IpfsNode, ctx context.Context) error {
-	rmed, err := gc.GC(ctx, n.Blockstore, n.Pinning)
+	kr, err := GarbageCollectAsync(n, ctx)
 	if err != nil {
 		return err
 	}
 
-	for {
-		select {
-		case _, ok := <-rmed:
-			if !ok {
-				return nil
-			}
-		case <-ctx.Done():
-			return ctx.Err()
-		}
+	// block until gc completes
+	for _ = range kr {
 	}
 
+	return nil
 }
 
 func GarbageCollectAsync(n *core.IpfsNode, ctx context.Context) (<-chan *KeyRemoved, error) {
-	rmed, err := gc.GC(ctx, n.Blockstore, n.Pinning)
+	// GC blocks from data blockstore
+	rmed, err := gc.GC(ctx, n.DataBlocks, n.Pinning)
+	if err != nil {
+		return nil, err
+	}
+
+	// GC blocks from state blockstore
+	ks := key.NewKeySet()
+	for _, k := range n.Pinning.InternalPins() {
+		ks.Add(k)
+	}
+
+	internal, err := gc.RunGC(ctx, n.StateBlocks, ks)
```

Comment:
this is the sort of thing i worried about, and it's starting to happen. this split is concerning at best.